src/plugins/flattenModifier.ts

```diff
@@ -52,8 +68,167 @@ export class FlattenModifier {
         copy[NodeCliHelper.CLI_FLATTEN] = src[NodeCliHelper.CLI_FLATTEN];
         return copy;
     }
+
+    private processSchema(setting: FlattenSetting): void {
+        this.session.model.schemas.objects?.forEach((schema) => {
+            this.flattenSchema(schema);
+        });
+
+        if (!setting.keepUnusedModel) {
+            this.removeUnusedModel();
+        }
+
+        for (const schema of values(this.session.model.schemas.objects)) {
+            if (schema.extensions) {
+                delete schema.extensions[isCurrentlyFlattening];
+                // don't want this until I have removed the unreferenced models.
+                // delete schema.extensions[hasBeenFlattened];
+                if (length(schema.extensions) === 0) {
+                    delete schema['extensions'];
+                }
+            }
+        }
+    }
+
+    private flattenSchema(schema: ObjectSchema): void {
+        const state = schema.extensions?.[isCurrentlyFlattening];
+
+        if (state === false) {
+            // already done.
+            return;
+        }
+
+        if (state === true) {
+            // in progress.
+            throw new Error(`Circular reference encountered during processing of x-ms-client flatten ('${schema.language.default.name}')`);
+        }
+
+        // hasn't started yet.
+        schema.extensions = schema.extensions || {};
+        schema.extensions[isCurrentlyFlattening] = true;
+
+        // ensure that parent schemas are done first -- this should remove 
+        // the problem when the order isn't just right.
+        for (const parent of values(schema.parents?.immediate)) {
+            if (isObjectSchema(parent)) {
+                this.flattenSchema(parent);
+            }
+        }
+
+        if (schema.properties) {
+            for (const { key: index, value: property } of items(schema.properties).toArray().reverse()) {
```

Comment:
>properties [](start = 71, length = 10)

there may be properties from parent, you can use the allproperties method from modelerfour library